### ISSUE TITLE
How to use nixpkgs_cc_configure?

### ISSUE THREAD

**Issue by prestonvanloon** [tweag/rules_nixpkgs#54]
I am seeing this error:
```
ERROR: in target '//external:cc_toolchain': no such package '@local_config_cc//': no such package '@nixpkgs_cc_toolchain//': Traceback (most recent call last):
        File "/home/preston/.cache/bazel/_bazel_preston/2bf51e8690558a83a0a0514ab422fd4b/external/io_tweag_rules_nixpkgs/nixpkgs/nixpkgs.bzl", line 122
                _executable_path(repository_ctx, "nix-build", extra_m...")
        File "/home/preston/.cache/bazel/_bazel_preston/2bf51e8690558a83a0a0514ab422fd4b/external/io_tweag_rules_nixpkgs/nixpkgs/nixpkgs.bzl", line 325, in _executable_path
                fail("Could not find the `{}` executa... ""))
Could not find the `nix-build` executable in PATH. See: https://nixos.org/nix/
INFO: Elapsed time: 0.258s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (1 packages loaded)
FAILED: Build did NOT complete successfully (1 packages loaded)
```

My workspace looks something like this: 

```python
http_archive(
    name = "io_tweag_rules_nixpkgs",
    strip_prefix = "rules_nixpkgs-f483fed8a83a4fd07ebc20aab9bbac2df11b256f",
    urls = ["https://github.com/tweag/rules_nixpkgs/archive/f483fed8a83a4fd07ebc20aab9bbac2df11b256f.tar.gz"],
)

load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl", "nixpkgs_cc_configure", "nixpkgs_git_repository", "nixpkgs_package")

nixpkgs_git_repository(
    name = "nixpkgs",
    revision = "17.09",  # Any tag or commit hash
    sha256 = "",  # optional sha to verify package integrity! (yes, I will after I can get it to work...)
)

nixpkgs_package(
    name = "hello",
    repositories = {"nixpkgs": "@nixpkgs//:default.nix"},
)

nixpkgs_cc_configure(repository = "@nixpkgs//:default.nix")

...
```
What am I missing to use `nixpkgs_cc_configure`? 
My goal here is to have more hermetic builds by not relying on the system gcc (inspired by @mboes [lightning talk at Bazelcon 2018](https://youtu.be/7-K_RmDasEg?t=2033)).

More info: 
```
bazel version
INFO: Invocation ID: fdb8124e-372e-43b0-a8e7-495d445988aa
Build label: 0.20.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Nov 30 14:39:01 2018 (1543588741)
Build timestamp: 1543588741
Build timestamp as int: 1543588741
```

**Comment by mboes** [comment]
> Could not find the `nix-build` executable in PATH. See: https://nixos.org/nix/

`rules_nixpkgs` requires this command to be available in your enviroment. See the provided URL for instructions on how to install Nix on your system. Think of `nix-build` kind of like the `docker build` command. Everything inside a docker image can be made reproducible, but you still need some (unspecified) version of the `docker` command available in your environment to do so.

**Comment by prestonvanloon** [comment]
@mboes shouldn't this be provided as part of bazel via these rules? 

Requiring yet-another-package-manager installed locally is a deal breaker for me.

**Comment by mboes** [comment]
> shouldn't this be provided as part of bazel via these rules?

If we could install Nix locally to some arbitrary path, yes. But we can't. Because if you want to benefit from the Nix cache (rather than building your GCC from scratch), then you'll need to install Nix in the same location on your system as everyone else is (i.e. `/nix/store`). 

> Requiring yet-another-package-manager installed locally is a deal breaker for me.

OK fair enough. I don't see it that way. Nix is just an alternative to Docker. You probably have APT or YUM on your system but you also have Docker, right? It would be nice if we could use Docker here but as I explain in the lightning talk you linked to, docker images a) don't compose well, and b) are often not very reproducible. So we use something else that's similar to it, but which does compose. Composability is important because it's what avoid resorting to coarse grained Docker layers. If we have fine-grained layers, then we can tell Bazel precisely which targets depend on which layers, hence minimizing unnecessary rebuilds.

**Comment by prestonvanloon** [comment]
> If we could install Nix locally to some arbitrary path, yes. But we can't. Because if you want to benefit from the Nix cache (rather than building your GCC from scratch), then you'll need to install Nix in the same location on your system as everyone else is (i.e. /nix/store).

Why couldn't you? The [rules_docker](https://github.com/bazelbuild/rules_docker) doesn't require docker to be installed to build the docker images. 

I'm sure nix is a great tool, but the a major benefit of bazel is to have a single build tool. So asking developers or users to install a second tool to build a project is just not going to work for me. 

(Yes, we assume that gcc, java, and python are installed locally as a minimum requirement to run bazel. However, these are not project specific dependencies).

**Comment by regnat** [comment]
> The rules_docker doesn't require docker to be installed to build the docker images.

That's true, but that's because `rules_docker` doesn't use docker at all (because docker images are just a tar.gz of the image content with a bunch of metadata, so you can build them without docker)

I think we could somehow run nix locally (provided we can get a self-contained installation of nix in bazel) using mount namespaces (nix has a native support for it for building the packages, even though that doesn't allow running executables which will still have hardcoded references to `/nix/store`).
That's definitely not trivial though because that probably would require us to customize the bazel sandboxes, and I have no idea whether bazel allows that.

**Comment by Profpatsch** [comment]
> even though that doesn't allow running executables which will still have hardcoded references to `/nix/store`

So what needs to happen: We wrap each executable in a script that passes through everything, but maps `/nix/store` to the nix store we construct somewhere on the user’s system (via `unshare(1)` for example). We call version the `nix` we extracted to the bazel cache with `--store /path/to/local/store`.

**Comment by zimbatm** [comment]
There is also https://github.com/matthewbauer/nix-bundle that uses similar tricks to produce a self-contained Nix executable. These could be distributed as artifacts of this repository.

One downside though is that Darwin doesn't have an equivalent for `unshare(1)` AFAIK. It might be worth looking at the Bazel implementation to see how they do it.

**Comment by prestonvanloon** [comment]
Thanks for considering and looking into this use case. It would be quite powerful to plug this into existing projects for more reliable builds.

**Comment by mboes** [comment]
@zimbatm Bazel use `sandbox-exec` on Darwin. There's a bog post about it: https://blog.bazel.build/2015/09/11/sandboxing.html.

**Comment by mboes** [comment]
It turns out that mapping local directories to global directories inside the sandbox is possible, using `--sandbox_add_mount_pair`. @prestonvanloon would requiring no more than this option being set in all projects' `.bazelrc` make rules_nixpkgs a viable option for you? Ideally, even this option wouldn't be necessary but it looks like this is as good as it's going to get without further improvements to Bazel itself.

**Comment by prestonvanloon** [comment]
I think that would be OK. I haven’t evaluated that flag for side effects, but it’s not a usability concern from a project perspective. Thanks!

**Comment by Profpatsch** [comment]
Closed in favor of #75
